### PR TITLE
adds selection persistence on sidebar tab switch

### LINKED ISSUE #13309: Keep the selected block in a selected state when switching the sidebar to Document
When a block is selected, it stays in the selected state also when operating on other parts of the user interface. It is possible to click on any part of the sidebar or top bar, and the block is still selected.

For example, it's possible to activate menus in the top bar, and the block is still selected:

screenshot 2019-01-13 at 14 22 56

This is actually a good thing: the interaction with the block is consistent and predictable. There are no assumption in the UI behavior and the block selected state is in the users full control.

There's one relevant case where this is no longer true: when clicking the sidebar "Document" tab, the selected block loses its selected state.

To better clarify, I've made a GIF:

block selected state

For mouse users:
it's maybe a minor annoyance but worth considering they have to move their mouse to the block, click into it to make it "selected" again, and move their mouse back to the sidebar again to be able to operate in the "block inspector" settings.

For keyboard users (including screen reader users, switch control users, etc.):
it's a serious breakage of keyboard navigation. Just imagine this flow:
- a block is selected
- users jump to the sidebar, either by using the `Ctrl + backtick` shortcut or using the navigation tools their assistive technology provides
- users switch the Sidebar to "Document" to explore its content or for any other reason
- they then switch the Sidebar to "Block"
- at that point, the block is not selected any longer and the inspector is empty
- to be able to operate again on the block settings, users have to tab multiple times through the whole editor interface to go back to the block and select it, which is a challenging task especially when there are many blocks
- then, they can finally jump again to the sidebar

As [reported by the accessibility team](https://make.wordpress.org/accessibility/2018/10/29/report-on-the-accessibility-status-of-gutenberg/), keyboard navigation between the content area and the sidebar is one of the main barriers for users with accessibility needs. Gutenberg should strive to make navigation through content as easy as possible.

Personally, I'm not sure there's even a good reason why the block should _not_ be selected when the sidebar switches to "Document". I'd tend to think the block selected state should be completely independent from the Sidebar state. 

Ideally, the block selected state should always be in full control of the user. The UI behavior shouldn't assume users intent.

I'd like to propose to keep the block selected when switching the sidebar to "Document". I've tried to have a look at how it works, and turned out it's a bit complicated as the sidebar switching logic is actually based on the presence of a selected block. I'd greatly appreciate if anyone with better knowledge of this implementation could have a look at this, as it would be a pretty good accessibility improvement.

### PR DESCRIPTION
**Pull request by draganescu**
## Description
Fixes #13309 

When a block is selected or multi selected switching the settings sidebar to document deselects everything but switching back to block selects everything back.

## How has this been tested?
There is an e2e test in the a11y suite.

## Screenshots 

restore-selection

## Types of changes
Non breaking changes.
Added a new action in the core/block-editor store, restoreSelectedBlock.
Updated the blockSelection reducer of the core/block-editor store.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate.

### CODE REVIEW

**Comment by draganescu** [comment]
@gziolo using the page.$ selection engine from puppeteer I couldn't get 'is-selected' no matter what I did, hence the eval solution in the test. Not sure why, but the selector was always null, even though visual inspection contradicted that.

The state update in this PR is generic enough, the only required implementation is to dispatch the new action whenever we want to return the previous selection after clearing.

File: packages/block-editor/src/store/actions.js
Comment (gziolo) [inline packages/block-editor/src/store/actions.js]: Please note that there is a slight difference between the function name and its description. It would more accurate to have:

```js
export function restoreBlockSelection() {
	return {
		type: 'RESTORE_BLOCK_SELECTION',
	};
}
```

I see it follows `clearSelectedBlock` so I'm not sure what's the best way to handle it as changing both would require to keep an alias for the existing method to keep it backward compatible ... 🤷‍♂ @youknowriad - I guess it's fine to keep it singular, right?

File: packages/block-editor/src/store/reducer.js
Comment (gziolo) [inline packages/block-editor/src/store/reducer.js]: How about using a simpler syntax like with `omit` from lodash:
```js
return {
				...BLOCK_SELECTION_INITIAL_STATE,
				previousSelection: omit( state, [ 'previousSelection' ] ),
};
```
It would simplify maintanance if the state shape changes in the future.

Can we also add a simple unit test which covers this action in the reducer?

File: packages/e2e-tests/specs/a11y.test.js
Comment (gziolo) [inline packages/e2e-tests/specs/a11y.test.js]: There is also this handy helper method which can simplify this code to:
```js
await pressKeyWithModifier( 'ctrlShift', '`' );
await pressKeyWithModifier( 'ctrlShift', '`' );
```

It would be nice to add a code comment for those who don't know this keyboard shortcut.

Just noting that it moves focus backwards.

File: packages/e2e-tests/specs/a11y.test.js
Comment (gziolo) [inline packages/e2e-tests/specs/a11y.test.js]: It might make it easier to follow this test if we add assertion whether the currently focused element is the Document tab button here and the Block tab button in the next step.

**Review by gziolo — CHANGES_REQUESTED** [review]
> The state update in this PR is generic enough, the only required implementation is to dispatch the new action whenever we want to return the previous selection after clearing.

Yes, it looks very clean in the current shape. I like this proposal 👍 
There is some more work required on this branch as noted in code comments. Travis isn't happy about the following as well: 

1. There are failing unit tests with these changes applied:
https://travis-ci.com/WordPress/gutenberg/jobs/199430106#L5455

I think it's expected that there are going to be actions where the previous selection will have to be cleared as well.

2. There are new public API methods added so you will have to run `npm run docs:build` to refresh docs.


3. There is one e2e test failing which looks completely unrelated:
https://travis-ci.com/WordPress/gutenberg/jobs/199430115#L1092

```
FAIL packages/e2e-tests/specs/editor-modes.test.js (11.732s)
  ● Editing modes (visual/HTML) › the code editor should unselect blocks and disable the inserter
    expect(received).not.toBeNull()
    Received: null
      112 | 		await page.click( '.edit-post-sidebar__panel-tab[data-label="Block"]' );
      113 | 		const noBlocksElement = await page.$( '.block-editor-block-inspector__no-blocks' );
    > 114 | 		expect( noBlocksElement ).not.toBeNull();
          | 		                              ^
      115 | 
      116 | 		// The inserter is disabled
      117 | 		const disabledInserter = await page.$( '.block-editor-inserter > button:disabled' );
      at toBeNull (specs/editor-modes.test.js:114:33)
      at tryCatch (../../node_modules/regenerator-runtime/runtime.js:45:40)
      at Generator.invoke [as _invoke] (../../node_modules/regenerator-runtime/runtime.js:271:22)
      at Generator.prototype.(anonymous function) [as next] (../../node_modules/regenerator-runtime/runtime.js:97:21)
      at asyncGeneratorStep (specs/editor-modes.test.js:11:103)
      at _next (specs/editor-modes.test.js:13:194)
```

File: packages/block-editor/src/store/reducer.js
Comment (youknowriad) [reply packages/block-editor/src/store/reducer.js]: Some questions:

 - When do we clear this cache? Do we keep the previous selection indefinitely?
 - What about the other actions that change the selection (SELECT_BLOCK, INSERT_BLOCK...), do we want to save the previous selection there as well?

Comment (gziolo) [reply packages/block-editor/src/store/reducer.js]: Yes, there might be an edge case based on the failing e2e test. I think when you have a selection and switch to the Code editor and then back to the Visual editor, it might try to restore the selection. It needs to be further investigated as it might break when you remove all blocks in the Code editor mode.

File: packages/block-editor/src/store/reducer.js
Comment (aduth) [inline packages/block-editor/src/store/reducer.js]: Do we need to spread? Or can we simply do:

```js
return state.previousSelection;
```

A spread operator is more costly, since it creates a new object, and performs a shallow clone of `previousSelection` into that new object.

File: packages/block-editor/src/store/actions.js
Comment (draganescu) [reply packages/block-editor/src/store/actions.js]: I like your suggestion about the new name, but I think it is better because it is memorable to stick with restoreSelectedBlock as it is clearly related to restoreSelectedBlock. So we either rename both (and create aliases) or we keep them like this.

Personally I don't like adding empty code with aliases, but that is only pure preference as I also value clarity in naming.

File: packages/block-editor/src/store/reducer.js
Comment (draganescu) [reply packages/block-editor/src/store/reducer.js]: @youknowriad yes we can use restoreSelectedBlock in many places but I suggest not in this PR. One example, undo and redo could also take selection into account and return focus: I have a focus in a Paragraph block and I use the inserter to add an Image, when I undo it would be nice to have the focus returned (or selection reapplied).

But I think they should be incremental code changes that store the previous selection where it is needed and then use restoreSelectedBlock. This PR is only about the inspector tho.

Indeed, I will investigate places where the previousSelection cache has to be cleared so it will not result in errors.

**Comment by draganescu** [comment]
fixed two tests, made a new one fail! yay! :)

**Comment by draganescu** [comment]
I had to add a way to clear previousSelection destructively, that is without saving previous selection, so that restoreBlock wouldn't find anything to restore when invoked. The case was needed when selecting text mode, and called restoreBlock it looked in the UI as if there were some blocks selected, even though none existed.

File: packages/block-editor/src/store/actions.js
Comment (draganescu) [inline packages/block-editor/src/store/actions.js]: we should make another action for the destructive way!

**Comment by draganescu** [comment]
we should wipe the selection when the blocks are intentionally unselected and write a test about that.

**Comment by draganescu** [comment]
I have added a wipeBlockSelection as a way to destructively clear block selection and called it on visual/text editor mode switch and also when all blocks are manually deselected.

File: packages/block-editor/src/store/actions.js
Comment (gziolo) [inline packages/block-editor/src/store/actions.js]: This type param is no longer user, it should be removed.

File: packages/e2e-tests/specs/a11y.test.js
Comment (gziolo) [inline packages/e2e-tests/specs/a11y.test.js]: To be precise, this checks whether RichText component is focused rather than any block. In case of Paragraph block it happens to be the only component used.

**Review by gziolo — COMMENTED** [review]
I will retest later today, code wise it looks good 👍

File: packages/e2e-tests/specs/a11y.test.js
Comment (draganescu) [reply packages/e2e-tests/specs/a11y.test.js]: I don't really understand what you mean :)

File: packages/block-editor/src/store/actions.js
Comment (draganescu) [reply packages/block-editor/src/store/actions.js]: Oups! Solved.

File: packages/e2e-tests/specs/a11y.test.js
Comment (gziolo) [reply packages/e2e-tests/specs/a11y.test.js]: This doesn't check whether Paragraph block is focused as the name implies: `isFocusedParagraphBlock`. It checks whether RichText component is focused. There are plenty of blocks which use RichText component(s) so this is confusing.

File: packages/block-editor/src/store/actions.js
Comment (gziolo) [inline packages/block-editor/src/store/actions.js]: > Returns an action object used in signalling that the block selection is cleared.

Both `clearSelectedBlock` and `wipeSelectedBlock` have the same description. It should be further clarified so developers could decide which one fits better to their use case.

File: packages/block-editor/src/store/reducer.js
Comment (gziolo) [inline packages/block-editor/src/store/reducer.js]: It would be great to cover `WIPE_SELECTED_BLOCK` and `RESTORE_SELECTED_BLOCK` reducer updates with unit tests.

**Review by gziolo — COMMENTED** [review]
This works great, I tested it with single and multiple blocks selected and it restores selections after I bring back Block settings sidebar. I tested it also with switching to the code editor and unselecting blocks after clicking on editor's canvas and it doesn't restore the selection as expected 👍 

There is still two comments to address before it gets merged.

File: packages/block-editor/src/store/reducer.js
Comment (draganescu) [reply packages/block-editor/src/store/reducer.js]: added the tests!

File: packages/e2e-tests/specs/a11y.test.js
Comment (draganescu) [reply packages/e2e-tests/specs/a11y.test.js]: updated the wording and the flow to match what is really happening.

**Comment by draganescu** [comment]
@gziolo everything should be up to date and conform to the requested changes :)

File: packages/block-editor/src/store/reducer.js
Comment (gziolo) [reply packages/block-editor/src/store/reducer.js]: Those are tests for actions, they don't test the reducer :)
Tests for actions aren't that helpful as those are simple wrappers on static objects.

**Review by gziolo — APPROVED** [review]
Thank you for addressing this issue. This is a really great improvement! Code looks great 💯

**Comment by aduth** [comment]
Coming to this late, but a few brief observations in passing:

- Why should changing between "Document" and "Blocks" have any impact on selection, either visually or in data? From what I can discern from the animation, it seems we still _visually_ show the blocks as deselected when using the "Document" tab. I'm unsure if this is what motivated some of the new implementation.
- It wasn't immediately obvious to me what the new terminology meant with "wipe" a selection. If the idea is that clearing a selection can be reverted, I wonder if we could have instead done this by some alternative of (a) using/modifying state history (e.g. [`withHistory`](https://github.com/WordPress/gutenberg/tree/master/packages/editor/src/utils/with-history)), (b) using [`redux-optimist`](https://github.com/ForbesLindesay/redux-optimist) (leveraged also in editor's `reducer.js`) to have this operate as a commit/revert operation, or (c) just as an argument of the existing `clearSelectedBlock` to reflect this effect.
   - Personally I think we should be wary to introduce new actions when existing ones can serve just as well with minor (sensible) revisions
   - Depending on the answer to my first point, I wonder if we'd need any of this at all
- I see a warning now when viewing the editor with `SCRIPT_DEBUG`. I assume this prop is being passed down somewhere it's not meant to be:

```
react-dom.165d5c53.js:500 Warning: React does not recognize the `wipeSelectedBlock` prop on a DOM element. If you intentionally want it to appear in the DOM as a custom attribute, spell it as lowercase `wipeselectedblock` instead. If you accidentally passed it from a parent component, remove it from the DOM element.
    in div (created by BlockSelectionClearer)
    in BlockSelectionClearer (created by WithDispatch(BlockSelectionClearer))
    in WithDispatch(BlockSelectionClearer)
    in Unknown (created by WithSelect(WithDispatch(BlockSelectionClearer)))
    in WithSelect(WithDispatch(BlockSelectionClearer)) (created by VisualEditor)
    in VisualEditor (created by Layout)
    in div (created by Layout)
    in div (created by FocusReturnProvider)
    in FocusReturnProvider (created by Layout)
    in Layout
    in Unknown (created by WithViewportMatch(Layout))
    in WithViewportMatch(Layout) (created by NavigateRegions(WithViewportMatch(Layout)))
    in div (created by NavigateRegions(WithViewportMatch(Layout)))
    in NavigateRegions(WithViewportMatch(Layout)) (created by WithDispatch(NavigateRegions(WithViewportMatch(Layout))))
    in WithDispatch(NavigateRegions(WithViewportMatch(Layout)))
    in Unknown (created by WithSelect(WithDispatch(NavigateRegions(WithViewportMatch(Layout)))))
    in WithSelect(WithDispatch(NavigateRegions(WithViewportMatch(Layout)))) (created by Editor)
    in ErrorBoundary (created by Editor)
    in div (created by DropZoneProvider)
    in DropZoneProvider (created by BlockEditorProvider)
    in SlotFillProvider (created by BlockEditorProvider)
    in BlockEditorProvider (created by WithDispatch(BlockEditorProvider))
    in WithDispatch(BlockEditorProvider)
    in Unknown (created by Context.Consumer)
    in WithRegistryProvider(WithDispatch(BlockEditorProvider)) (created by EditorProvider)
    in EditorProvider (created by WithDispatch(EditorProvider))
    in WithDispatch(EditorProvider)
    in Unknown (created by WithSelect(WithDispatch(EditorProvider)))
    in WithSelect(WithDispatch(EditorProvider)) (created by Editor)
    in StrictMode (created by Editor)
    in Editor
    in Unknown (created by WithSelect(Editor))
    in WithSelect(Editor)
```

**Comment by draganescu** [comment]
oh boy :) so many things. 

First the decision to deselect blocks when using the document inspector is an older [design decision](https://github.com/WordPress/gutenberg/issues/13309#issuecomment-458452168) a while ago .

Second, I had no idea about `redux-optimist`, I was trying to solve the issue using a straight forward approach: make the thing reversible :O) by storing it in a place.

Indeed, clear vs wipe, maybe it's not the best :) Naming is hard. Initially I wanted to pass an object as an argument to `clearSelectedBlock` like `{ destructively: true }` just to make it clearer.

I will create an issue for the error on `SCRIPT_DEBUG`. I think this should default to ON in the dev Gutenberg docker image?

**Comment by aduth** [comment]
> First the decision to deselect blocks when using the document inspector is an older [design decision](https://github.com/WordPress/gutenberg/issues/13309#issuecomment-458452168) a while ago .

Thanks for sharing the context. I hadn't seen @jasmussen 's comment. It helps frame this pair of tabs as something of being mutually exclusive to there being / not being a selection of blocks, which I wasn't aware of.

Aside from this point, one of the motivators in my prior comment is in limiting what we add to our own maintenance burden if existing concepts could apply. `withHistory` and `redux-optimist` may or may not work well for this, but conceptually it seemed like the idea proposed here aligned with some behavior of a reversal of some prior action (the deselection). Ultimately, going this route may prove to add more complexity than what you've proposed, so I don't mean for it to come across as a challenge, but rather as constructive contemplation. Similarly, if we could account for this behavior in the existing `clearSelectedBlock` or by just changing the UI visually (while retaining the selection state), there could be some middle ground. I was prompted to make my comment because it wasn't entirely clear to me what was meant by "wipe", at least in contrast with the existing "clear" mechanic. I expect a future maintainer may have similar difficulty in discerning them.

> I will create an issue for the error on `SCRIPT_DEBUG`. I think this should default to ON in the dev Gutenberg docker image?

I noticed after making my comment that @afercia had commented similarly at https://github.com/WordPress/gutenberg/issues/13309#issuecomment-499097187 . Since this issue has been reopened, it may be enough to use that one for tracking?

I thought we did have `SCRIPT_DEBUG` as the default for the Docker development environment (#14371). It isn't yet enabled for end-to-end testing, since there are a number of existing React warnings (fewer as of #11360) which can trip the global error capturing. The task to enable this is tracked at #14845.

**Comment by jasmussen** [comment]
> Thanks for sharing the context. I hadn't seen @jasmussen 's comment. It helps frame this pair of tabs as something of being mutually exclusive to there being / not being a selection of blocks, which I wasn't aware of.

Currently you can customize the document loaded, and you can customize each block in that document. One of the things that may be explored in the future, is to allow you to customize also the site itself (i.e. Site | Document | Block) as tabs. The argument made in the past, and thanks for considering it, is that by leveraging breadcrumbs instead of tabs, we make the hierarchy clear (Site → Document → Block). It also has the benefit that breadcrumbs can help you indicate this hierarchy even further (Site → Document → Columns → Column → Quote → Paragraph), and finally that both in cases of translation length and for the longer breadcrumb, it can wrap to two lines, whereas tabs can't. But at the same time, it is critical to such breadcrumbs that they are directly tied to what is selected. If you have a block selected and click the "Site" breadcrumb, that's what should get selected.

**Comment by youknowriad** [comment]
Based on the fact that there are some technical uncertainties and that the original issue is not really fixed by that PR, I'm wondering if we should revert this for this release and give this more time to mature and address these concerns.

**Pull request merged**
